### PR TITLE
set ip/ipv6 next_hop 'native' option

### PR DESCRIPTION
**Pull request by aida-shumburo**
##### SUMMARY
Set up the 'native' option for both IP_next_hop and IPv6_next_hop.

##### GitHub Issues
"N/A".

| GitHub Issue # |
| -------------- |
| |

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
sonic_route_maps

##### Ansible Report:
[add_native_option.pdf](https://github.com/user-attachments/files/16732602/add_native_option.pdf)

##### Raw Regression Test Output:
[set_native_option.zip](https://github.com/user-attachments/files/16732727/set_native_option.zip)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] I ran different ansible playbooks to test that the new 'native' option value is being merged, replaced, deleted, and overriden the correct way for both ip_next_hop and ipv6_next_hop
- [x] Ran a regression test (attached above)

### CODE REVIEW

File: plugins/modules/sonic_route_maps.py [hunk 1]
```diff
@@ -313,6 +313,15 @@
               - IPv4 next hop address to set into a matching route in the
               - dotted decimal format A.B.C.D
             type: str
+          ip_next_hop_options:
```
Comment (kerry-meyer) [inline plugins/modules/sonic_route_maps.py:316]: After some discussion and further consideration on this point, I would like to recommend that adding the 'native' option for IPv4 next hops be done by changing the current "ip_next_hop" option into a dictionary (similar to what is already implemented for IPv6) and locating the 'native' option in that dictionary, along with the current 'str' option, with the key/name of 'address'.

This same change would also need be done in the argspec and in the route map "model" (in the "models" repo).

While this is a "breaking" change and will need to be deferred to the 3.0.0 release, it will make the module/argspec/model more symmetrical for the two address families and easier for customers to understand and use versus splitting the IPv4 "next hop" options into two different option groups.

File: plugins/modules/sonic_route_maps.py [hunk 2]
```diff
@@ -310,9 +310,16 @@
...
+              - IPv4 next hop address attributes to set into a matching route
+            type: dict
+            suboptions:
+              address:
+                - IPv4 next hop address to set into a matching route in the
+                - dotted decimal format A.B.C.D
+              type: str
+              native:
+                description: Set native or underlay nexthop
+                type: bool
```
Comment (kerry-meyer) [inline plugins/modules/sonic_route_maps.py:322]:
```suggestion
              - IPv4 next hop address attributes to set into a matching route
            type: dict
            suboptions:
              address:
                description:
                  - IPv4 next hop address to set into a matching route in the
                  - dotted decimal format A.B.C.D
                type: str
              native:
                description: Set native or underlay nexthop
                type: bool
```

File: plugins/modules/sonic_route_maps.py [hunk 3]
```diff
@@ -1171,6 +1181,8 @@
              soo:
                - "10.73.14.9:78"
            ip_next_hop: 10.48.16.18
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/modules/sonic_route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 4]
```diff
@@ -2325,6 +2392,38 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
...
+                                cfg_set_top['ip_next_hop'][member_key]
+                            request = {'path': set_uri[member_key], 'method': DELETE}
+                            dict_delete_requests.append(request)
+            command['set'].pop('ip_next_hop')
+            if ip_next_hop_deleted_members:
+                command['set']['ip_next_hop'] = ip_next_hop_deleted_members
+
         # If the "replaced" command set includes ipv6_next_hop attributes that
         # differ from the currently configured attributes, delete
         # ipv6_next_hop configuration, if it exists, for any ipv6_next hop
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 5]
```diff
@@ -2159,6 +2196,36 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
...
+                for ip_next_hop_key in cfg_ip_next_hop_key_set.difference(cmd_ip_next_hop_key_set):
+                    ip_next_hop_deleted_members[ip_next_hop_key] = \
+                        cfg_set_top['ip_next_hop'][ip_next_hop_key]
+                    dict_delete_requests.append(request)
+
+                if ip_next_hop_deleted_members:
+                    # Update the list of deleted ip_next_hop attributes in the "command" dict.
+                    if not cmd_set_top.get('ip_next_hop'):
+                        command['set']['ip_next_hop'] = {}
+
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py:2226]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 6]
```diff
@@ -2325,6 +2392,38 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
                     }
                     dict_delete_requests.append(request)
 
+
+        # differ from the currently configured attributes, delete
+        # attributes that are not specified in the received command.
+        # If the "replaced" command set includes ip_next_hop attributes that
+        # differ from the currently configured attributes, delete
+        # ip_next_hop configuration, if it exists, for any ipv6_next hop
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 7]
```diff
@@ -2325,6 +2392,38 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
                     }
                     dict_delete_requests.append(request)
 
+
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 8]
```diff
@@ -2159,6 +2196,36 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
...
                     dict_delete_requests.append(request)
 
+            # Check for deletion of ip_next_hop attributes. Delete the attributes
+            # As an optimization, avoid deleting attributes that will be replaced
+            # by the received command.
+            # by the received command.
+            ip_next_hop_deleted_members = {}
+            if 'ip_next_hop' not in cfg_set_top:
+                if command['set'].get('ip_next_hop'):
+                    if command['set'] is None:
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py:2205]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 9]
```diff
@@ -2159,6 +2196,36 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
                     }
                     dict_delete_requests.append(request)
 
+            # Check for deletion of ip_next_hop attributes. Delete the attributes
+            # As an optimization, avoid deleting attributes that will be replaced
+            # by the received command.
+            # by the received command.
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 10]
```diff
@@ -1944,10 +1976,15 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
             'comm_list_delete': bgp_set_delete_req_base + 'set-community-delete',
             'community': bgp_set_delete_req_base + 'set-community',
             'extcommunity': bgp_set_delete_req_base + 'set-ext-community',
-            'ip_next_hop': bgp_set_delete_req_base + 'config/set-next-hop',
+            'extcommunity': bgp_set_delete_req_base + 'set-ext-community',
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Removed

File: plugins/modules/sonic_route_maps.py (hunk 2 above)
Comment (aida-shumburo) [reply plugins/modules/sonic_route_maps.py]: updated

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 11]
```diff
@@ -2325,9 +2392,38 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
                     }
                     dict_delete_requests.append(request)
 
+        # If the "replaced" command set includes ip_next_hop attributes that
+        # differ from the currently configured attributes, delete
+        # ip_next_hop configuration, if it exists, for any ip_next hop
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Fixed.

File: tests/unit/modules/network/sonic/fixtures/sonic_route_maps.yaml [hunk 12]
```diff
@@ -3658,7 +3857,8 @@ deleted_01:
               - "30:40"
             soo:
               - 10.73.14.9:78
-          ip_next_hop: 10.48.16.18
+          ip_next_hop:
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply tests/unit/modules/network/sonic/fixtures/sonic_route_maps.yaml:3860]: I noticed the delete_05 is the same test as delete_04 so I am updating delete_05 test to test the 'native' attribute for both ip_next_hop and ipv6_next_hop.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 13]
```diff
@@ -1461,18 +1471,42 @@ def get_route_map_delete_set_bgp_cfg(self, command, set_both_keys, cmd_rmap_have
...
-                'prefer_global': 'set-ipv6-next-hop-prefer-global'
+                'native': 'openconfig-bgp-policy-ext:set-ipv6-next-hop-native',
+                'prefer_global': 'set-ipv6-next-hop-prefer-global',
             }
             for ipv6_next_hop_key in ipv6_next_hop_rest_names:
                 if cmd_set_top['ipv6_next_hop'].get(ipv6_next_hop_key) is not None:
                     if (cmd_set_top['ipv6_next_hop'][ipv6_next_hop_key] ==
                             cfg_set_top['ipv6_next_hop'].get(ipv6_next_hop_key)):
                         delete_bgp_attrs.append(ipv6_next_hop_rest_names[ipv6_next_hop_key])
+
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Got rid of the extra blank line.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml [hunk 14]
```diff
@@ -299,9 +310,9 @@ tests:
               - "30:40"
             soo:
               - 10.73.14.9:78
-          ip_next_hop: 10.48.16.18
+          ip_next_hop: 
+            native: true
           ipv6_next_hop:
-            global_addr: 30::30
```
Comment (kerry-meyer) [inline tests/regression/roles/sonic_route_maps/defaults/main.yml:304]: Why is this line deleted?

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 15]
```diff
@@ -2159,6 +2194,37 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
...
+                cfg_ip_next_hop_key_set = set(cfg_set_top['ip_next_hop'].keys())
+                cmd_ip_next_hop_key_set = ([])
+                if cmd_set_top.get('ip_next_hop'):
+                    cmd_ip_next_hop_key_set = set(cmd_set_top['ip_next_hop'].keys())
+
+                set_uri = set_uri_attr['ip_next_hop']
+                for ip_next_hop_key in cfg_ip_next_hop_key_set.difference(cmd_ip_next_hop_key_set):
+                    ip_next_hop_deleted_members[ip_next_hop_key] = \
+                        cfg_set_top['ip_next_hop'][ip_next_hop_key]
+                    dict_delete_requests.append(request)
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Updated.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 16]
```diff
@@ -2159,6 +2194,37 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
                     }
                     dict_delete_requests.append(request)
 
+            # Check for deletion of ip_next_hop attributes. Delete the attributes
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Deleted the rest of the line.

File: plugins/modules/sonic_route_maps.py
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply plugins/modules/sonic_route_maps.py:1]: Updated the instances of "ip_next_hop" and "ipv6_next_hop"

File: changelogs/fragments/421-set-new-native-option-ipv-ipv6-next-hop.yaml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - sonic_route_maps - set ip/ipv6 next_hop 'native' option. This introduces a breaking change. (https://github.com/ansible-collections/dellemc.enterprise_sonic/pull/421)
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply changelogs/fragments/421-set-new-native-option-ipv-ipv6-next-hop.yaml]: Made the change.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml (hunk 14 above)
Comment (aida-shumburo) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml:304]: Added that line back.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml [hunk 18]
```diff
@@ -299,9 +310,9 @@ tests:
               - "30:40"
             soo:
               - 10.73.14.9:78
-          ip_next_hop: 10.48.16.18
+          ip_next_hop: 
+            native: true
```
[the remark this answers is not in the document]
Comment (aida-shumburo) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml:306]: Deleted that instance of ip_next_hop address.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py [hunk 19]
```diff
@@ -2159,6 +2194,38 @@ def get_delete_route_map_replaced_set_groupings(self, command, cmd_rmap_have,
...
+                        cfg_set_top['ip_next_hop'][ip_next_hop_key]
+                    request = {'path': set_uri[ip_next_hop_key], 'method': DELETE}
+                    dict_delete_requests.append(request)
+
+                if ip_next_hop_deleted_members:
+                    # Update the list of deleted ip_next_hop attributes in the "command" dict.
+                    if not cmd_set_top.get('ip_next_hop'):
+                        command['set']['ip_next_hop'] = {}
+
+                command['set']['ip_next_hop'] = ip_next_hop_deleted_members
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/route_maps/route_maps.py:2227]: This should only be done when the "if ip_next_hop_deleted_member" condition is true, so it should be indented to place it in the 'if' block.

```suggestion
                    command['set']['ip_next_hop'] = ip_next_hop_deleted_members
```

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Verified.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py (hunk 19 above)
Comment (aida-shumburo) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Updated.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Verified.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml [hunk 20]
```diff
@@ -78,6 +78,14 @@ tests:
...
           source_vrf: Vrf1
           tag: 7284
+        set:
+          ip_next_hop: 
+            address: 10.48.16.18
+            native: true
+          ipv6_next_hop:
+            global_addr: 30::30
+            prefer_global: true
+            native: true
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml]: Verified.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml:306]: It looks like this update got missed.

The address is still not being deleted.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml (hunk 14 above)
Comment (kerry-meyer) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml:304]: Verified.

File: plugins/module_utils/network/sonic/config/route_maps/route_maps.py (hunk 19 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/route_maps/route_maps.py]: Verified.

File: plugins/modules/sonic_route_maps.py
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_route_maps.py:1]: In the "after" state displays, the "show running-configuration" entries for the "native" options (ip and ipv6) should be shown after configuring them.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply tests/regression/roles/sonic_route_maps/defaults/main.yml:307]: Thanks for adding this option for deletion and also for adding "native" for the IPv6 next hop attribute deletion testing.

Resolved.

File: plugins/modules/sonic_route_maps.py
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_route_maps.py:1]: The "after" state displays have been updated as requested.

Resolved.

File: tests/regression/roles/sonic_route_maps/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_route_maps/defaults/main.yml:1]: All the test results look good for what was tested, but there is a missing test case that I found in looking at the results:

If 'native' options are set for ip/ipv6 next hops along with other next_hop options, a 'replaced' playbook that specifies only the non-native options should cause deletion of the 'native' options. I checked the code for this when I was reviewing, and it looked fine. But the regression test should test it.

To do that:

- For "test_case_09_merged_restore_rm1_80_set", add setting of the 'native' option for route map rm1 (seq 80) for the set ipv6_next_hop dictionary.
- For "test_case_11_replaced_dict_and_list",  add setting of the rm1/80 ip_next_hop address to a changed value and do not specify anything for 'native'.

Doing those two changes will test deletion of the 'native' option for the ip and ipv6 "replaced dict" cases.
Sorry for this additional change request, but since all current test results are fine, I think this should be a fairly quick and easy change to make and verify.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
All current proposed code changes and test results look good, but there is one additional test case that should be added to verify a case that is not currently being verified:

(I posted a corresponding comment for the tests/regression/roles/sonic_route_maps/defaults/main.yml file.)

**********

If 'native' options are set for ip/ipv6 next hops along with other next_hop options, a 'replaced' playbook that specifies only the non-native options should cause deletion of the 'native' options. I checked the code for this when I was reviewing, and it looked fine. But the regression test should test it.

To do that:

- For "test_case_09_merged_restore_rm1_80_set", add setting of the 'native' option for route map rm1 (seq 80) for the set ipv6_next_hop dictionary.
- For "test_case_11_replaced_dict_and_list",  add setting of the rm1/80 ip_next_hop address to a changed value and do not specify anything for 'native'.

Doing those two changes will test deletion of the 'native' option for the ip and ipv6 "replaced dict" cases.
Sorry for this additional change request, but since all current test results are fine, I think this should be a fairly quick and easy change to make and verify.

**Review by kerry-meyer — APPROVED** [review]
The current change set and test definitions and all of the corresponding test results look good.

Thank you for providing this new functionality for our resource module collection.

Approved.
